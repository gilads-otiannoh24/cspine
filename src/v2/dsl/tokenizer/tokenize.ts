import { splitArgs } from "../../utils/splitArgs";

export type Token =
  | { type: "command_group"; value: string }
  | { type: "command"; value: string }
  | { type: "command_type"; value: "call" | "normal" | "scope_use" }
  | { type: "reference"; value: string }
  | { type: "operator"; value: string }
  | { type: "arrow" }
  | { type: "cast"; value: string }
  | { type: "literal"; value: string }
  | { type: "pipe" }
  | { type: "positional_arg"; value: string }
  | {
      type: "named_arg";
      key: string;
      value: string;
      escaped: boolean;
      otherValues: string[];
    }
  | { type: "paren_open" }
  | { type: "paren_close" }
  | { type: "comma" }
  | { type: "semicolon" }
  | { type: "eof" };

export function tokenize(input: string, log: boolean = false): Token[] {
  const tokens: Token[] = [];
  const segments = input.split(/(?<=\))\s*;|;\s*/).filter(Boolean);

  if (log) console.log("Segments: ", segments);

  let i = 0;
  for (const segment of segments) {
    const type = getCommandType(segment);
    tokens.push({ type: "command_type", value: type });

    const regex =
      /@use\.[a-zA-Z_][a-zA-Z0-9_]*|\s+|(['"])(.*?)\1(?:\([a-zA-Z_][a-zA-Z0-9_]*\))?|"(.*?)"|\d+\([a-zA-Z_][a-zA-Z0-9_]*\)|[\$a-zA-Z_][\$a-zA-Z0-9_\.]*\([a-zA-Z_][a-zA-Z0-9_]*\)|[\$a-zA-Z_][\$a-zA-Z0-9_]*(?:\.[\$a-zA-Z_][\$a-zA-Z0-9_]*)?:[\$a-zA-Z_][\$a-zA-Z0-9_\.]*|->|\d+|true|false|\([a-zA-Z_][a-zA-Z0-9_]*\)|[+\-*/%]?=|\(|\)|,|;|\||[\$a-zA-Z_][\$a-zA-Z0-9_\.]*/g;

    let seg = segment + ";";
    let match;
    while ((match = regex.exec(seg))) {
      const text = match[0].trim();
      if (!text) continue;
      // support for @use.command
      else if (/^@use\.[a-zA-Z_][a-zA-Z0-9_]*$/.test(text)) {
        const commandName = text.slice("@use.".length);
        tokens.push({ type: "command", value: commandName });
      }

      if (
        /^([a-zA-Z_]+\.)?[a-zA-Z_]+:[\$a-zA-Z_][\$a-zA-Z0-9_\.]*$/.test(text)
      ) {
        const [left, ref] = text.split(":");
        const parts = left.split(".");

        if (parts.length === 2) {
          tokens.push({ type: "command_group", value: parts[0] });
          tokens.push({ type: "command", value: parts[1] });
        } else {
          tokens.push({ type: "command", value: parts[0] });
        }

        tokens.push({ type: "reference", value: ref });
      }

      // Casted number literal: 45(number)
      else if (/^\d+\([a-zA-Z_][a-zA-Z0-9_]*\)$/.test(text)) {
        const val = text.substring(0, text.indexOf("("));
        const cast = text.slice(text.indexOf("(") + 1, -1);
        tokens.push({ type: "literal", value: val });
        tokens.push({ type: "cast", value: cast });
      }

      // Quoted strings with optional cast: 'hello'(string) or "hello"(string)
      else if (/^(['"])(.*?)\1(?:\(([a-zA-Z_][a-zA-Z0-9_]*)\))?$/.test(text)) {
        const match = /^(['"])(.*?)\1(?:\(([a-zA-Z_][a-zA-Z0-9_]*)\))?$/.exec(
          text
        );
        if (match) {
          const value = match[2]; // inner quoted string
          const cast = match[3]; // optional cast
          tokens.push({ type: "literal", value });
          if (cast) tokens.push({ type: "cast", value: cast });
        }
      }

      // Plain quoted string literal: 'hello' or "hello"
      else if (/^(['"])(.*?)\1$/.test(text)) {
        const match = /^(['"])(.*?)\1$/.exec(text);
        if (match) {
          const [, , strVal] = match;
          tokens.push({ type: "literal", value: strVal });
        }
      }

      // Casted reference: user.id(number)
      else if (
        /^[\$a-zA-Z_][\$a-zA-Z0-9_\.]*\([\$a-zA-Z_][\$a-zA-Z0-9_]*\)$/.test(
          text
        )
      ) {
        const val = text.substring(0, text.indexOf("("));
        const cast = text.slice(text.indexOf("(") + 1, -1);

        const normalized = val.toLowerCase();

        if (normalized === "true" || normalized === "false") {
          tokens.push({ type: "literal", value: val });
          tokens.push({ type: "cast", value: cast });
        } else {
          tokens.push({ type: "reference", value: val });
          tokens.push({ type: "cast", value: cast });
        }
      }

      // Standalone cast
      else if (/^\([a-zA-Z_][a-zA-Z0-9_]*\)$/.test(text)) {
        tokens.push({ type: "cast", value: text.slice(1, -1) });
      }

      // Quoted strings
      else if (/^(['"])(.*?)\1$/.test(text)) {
        const quoteMatch = /^(['"])(.*?)\1$/.exec(text);
        if (quoteMatch) {
          const value = quoteMatch[2]; // inner content
          tokens.push({ type: "literal", value });
        }
      }
      // Arrow
      else if (text === "->") {
        tokens.push({ type: "arrow" });
      }

      // Pipe
      else if (text === "|") {
        tokens.push({ type: "pipe" });

        const argText = seg.slice(regex.lastIndex).trimStart();
        const endOfArgs = argText.search(/[;]/);
        const argsBlock =
          endOfArgs === -1 ? argText : argText.slice(0, endOfArgs);

        const args = splitArgs(argsBlock);

        for (const arg of args) {
          if (!arg) continue;

          const namedArgMatch = /^\\?([a-zA-Z0-9_]+)=(['"]?)(.*?)\2$/.exec(
            arg.trim()
          );

          if (namedArgMatch) {
            const [fullArg, rawKey, , rawValue] = namedArgMatch;

            const escaped = fullArg.startsWith("\\");
            const key = rawKey.startsWith("\\") ? rawKey.slice(1) : rawKey;
            const [value, ...otherValues] = rawValue.trim().split(",");
            tokens.push({
              type: "named_arg",
              key,
              value,
              otherValues,
              escaped,
            });
          } else {
            const value = arg.trim().replace(/^['"]|['"]$/g, "");
            tokens.push({ type: "positional_arg", value });
          }
        }

        regex.lastIndex += argsBlock.length;
      }

      // Punctuation
      else if (text === ",") {
        tokens.push({ type: "comma" });
      } else if (text === ";") {
        tokens.push({ type: "semicolon" });
      } else if (text === "(") {
        tokens.push({ type: "paren_open" });
      } else if (text === ")") {
        tokens.push({ type: "paren_close" });
      }

      // Operator
      else if (/^[+\-*/%]?=$/.test(text)) {
        tokens.push({ type: "operator", value: text });
      }

      // true and false as literals
      else if (text === "true" || text === "false") {
        tokens.push({ type: "literal", value: text });
      }

      // other literals (numbers)
      else if (/^\d+$/.test(text)) {
        tokens.push({ type: "literal", value: text });
      }

      // Fallback: references
      else if (
        /^[\$a-zA-Z_][\$a-zA-Z0-9_]*(\.[\$a-zA-Z_][\$a-zA-Z0-9_]*)*$/.test(text)
      ) {
        tokens.push({ type: "reference", value: text });
      }
    }

    i++;
  }

  tokens.push({ type: "eof" });
  return tokens;
}

function getCommandType(segment: string): "call" | "scope_use" | "normal" {
  const trimmed = segment.trim();
  const head = trimmed.split(/[|;]/)[0].trim(); // get part before args

  if (
    /^[\$a-zA-Z_][\$a-zA-Z0-9_]*(\.[\$a-zA-Z_][\$a-zA-Z0-9_]*)*\s*\(/.test(head)
  )
    return "call";

  if (/^@use\.[a-zA-Z_][a-zA-Z0-9_]*$/.test(head)) return "scope_use";

  return "normal";
}
