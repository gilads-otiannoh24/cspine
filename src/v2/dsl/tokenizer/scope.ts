// tokenizer/tokenizeScope.ts

import { Token, tokenize } from "./tokenize";

export type ScopeItem = {
  name: string;
  tokens: Token[];
};

export interface ScopeToken {
  commands: ScopeItem[];
  args: ScopeItem[];
  groups: ScopeItem[];
}

type CommandBlock = "commands" | "args" | "groups";

export function tokenizeScope(input: string, log: boolean = false): ScopeToken {
  const lines = input
    .split(/\n+/)
    .map((l) => l.trim())
    .filter((line) => line.length && !line.startsWith("//"));

  let currentBlock: CommandBlock | null = null;
  const tokens: ScopeToken = {
    commands: [],
    args: [],
    groups: [],
  };

  for (const line of lines) {
    if (/^(commands|args|groups):$/.test(line)) {
      currentBlock = line.replace(":", "") as CommandBlock;
      continue;
    }

    if (!currentBlock) continue;

    const [left, right] = line.split(":=").map((s) => s.trim());
    if (!left || !right) continue;

    let tokenList: Token[] = [];

    if (currentBlock === "commands") {
      tokenList = tokenize(right);
    } else {
      tokenList = tokenizeArgs(right);
    }

    tokens[currentBlock].push({ name: left, tokens: tokenList });
  }

  return tokens;
}

export function tokenizeArgs(input: string): Token[] {
  const result: Token[] = [];
  const argRegex =
    /\\?([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*(\".*?\"|\'.*?\'|\d+|true|false)/g;

  let match;
  while ((match = argRegex.exec(input)) !== null) {
    const [fullArg, key, rawValue] = match;
    const unquoted = rawValue.replace(/^['"]|['"]$/g, "");
    const [value, ...otherValues] = rawValue.trim().split(",");

    let val =
      value === "true"
        ? true
        : value === "false"
        ? false
        : /^\d+$/.test(value)
        ? Number(value)
        : unquoted;

    result.push({
      type: "named_arg",
      key,
      value: val as string,
      escaped: fullArg.startsWith("\\"),
      otherValues: otherValues,
    });
  }

  return result;
}
