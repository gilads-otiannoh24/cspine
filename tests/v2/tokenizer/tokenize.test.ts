import { tokenize } from "@/v2/dsl/tokenizer/tokenize";
import { describe, it, expect } from "vitest";

describe("Tokenizer", () => {
  it("should tokenize basic set command with literal", () => {
    const tokens = tokenize(`set:x->10`);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize casted literals and references", () => {
    const tokens = tokenize(`state.set:x->45(number);set:y->user.id(string)`);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize booleans and casts", () => {
    const tokens = tokenize(
      `set:ready->true(bool)|'user',option='foo,operator,name.id';set:flag->false(bool)`
    );
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize simple function call with no cast", () => {
    const tokens = tokenize(`log("debug")`);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize function call with mixed arguments", () => {
    const tokens = tokenize(
      `alert("Welcome"(string), user.id(string), 3000(number))`
    );
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize multiple calls separated by semicolon", () => {
    const tokens = tokenize(
      `show("Saved!"(string));refresh();log("done"(string))`
    );
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize reference chaining with @scope", () => {
    const tokens = tokenize(`set:@this.user.age->30(number)`);
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize @use scope reference", () => {
    const tokens = tokenize(
      `@use.set|silent=true;alert("Welcome"(string), user.id(string), 3000(number))`
    );
    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access as a target", () => {
    const tokens = tokenize(`state.set:user.name->name[0](string)`);

    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access (of type string) as a target", () => {
    const tokens = tokenize(`state.set:user.name->name['id'](string)`);

    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access (of type string without quotes) as a target", () => {
    const tokens = tokenize(`state.set:user.name->name[id](string)`);

    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access as a reference", () => {
    const tokens = tokenize(`state.set:user.name[0]->name`);

    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access (of type string) as a reference", () => {
    const tokens = tokenize(`state.set:user.name['id']->name`);

    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access (of type string without quotes) as a reference", () => {
    const tokens = tokenize(`state.set:user.name[id]->name`);

    expect(tokens).toMatchSnapshot();
  });

  it("should tokenize an array access (with deep nesting) as a reference", () => {
    const tokens = tokenize(`state.set:user.name['id'][0].friends[0].id->name`);

    expect(tokens).toMatchSnapshot();
  });

  it("should call function with the arguments (without dot operator) as reference unless quoted", () => {
    const tokens = tokenize(`fn(name)`, true);

    expect(tokens).toMatchSnapshot();
  });

  it("should call function with the arguments (without dot operator) as reference unless quoted", () => {
    const tokens = tokenize(`fn(name(string))`);

    expect(tokens).toMatchSnapshot();
  });
});
