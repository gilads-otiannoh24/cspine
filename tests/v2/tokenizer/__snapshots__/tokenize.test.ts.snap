// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tokenizer > should tokenize @use scope refenrence 1`] = `
[
  {
    "type": "command_type",
    "value": "scope_use",
  },
  {
    "type": "command",
    "value": "set",
  },
  {
    "type": "pipe",
  },
  {
    "escaped": false,
    "key": "silent",
    "otherValues": [],
    "type": "named_arg",
    "value": "true",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "command_type",
    "value": "call",
  },
  {
    "type": "reference",
    "value": "alert",
  },
  {
    "type": "paren_open",
  },
  {
    "type": "literal",
    "value": "Welcome",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "comma",
  },
  {
    "type": "reference",
    "value": "user.id",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "comma",
  },
  {
    "type": "literal",
    "value": "3000",
  },
  {
    "type": "cast",
    "value": "number",
  },
  {
    "type": "paren_close",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize basic set command with literal 1`] = `
[
  {
    "type": "command_type",
    "value": "normal",
  },
  {
    "type": "command",
    "value": "set",
  },
  {
    "type": "reference",
    "value": "x",
  },
  {
    "type": "arrow",
  },
  {
    "type": "literal",
    "value": "10",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize booleans and casts 1`] = `
[
  {
    "type": "command_type",
    "value": "normal",
  },
  {
    "type": "command",
    "value": "set",
  },
  {
    "type": "reference",
    "value": "ready",
  },
  {
    "type": "arrow",
  },
  {
    "type": "literal",
    "value": "true",
  },
  {
    "type": "cast",
    "value": "bool",
  },
  {
    "type": "pipe",
  },
  {
    "type": "positional_arg",
    "value": "user",
  },
  {
    "escaped": false,
    "key": "option",
    "otherValues": [
      "operator",
      "name.id",
    ],
    "type": "named_arg",
    "value": "foo",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "command_type",
    "value": "normal",
  },
  {
    "type": "command",
    "value": "set",
  },
  {
    "type": "reference",
    "value": "flag",
  },
  {
    "type": "arrow",
  },
  {
    "type": "literal",
    "value": "false",
  },
  {
    "type": "cast",
    "value": "bool",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize casted literals and references 1`] = `
[
  {
    "type": "command_type",
    "value": "normal",
  },
  {
    "type": "command_group",
    "value": "state",
  },
  {
    "type": "command",
    "value": "set",
  },
  {
    "type": "reference",
    "value": "x",
  },
  {
    "type": "arrow",
  },
  {
    "type": "literal",
    "value": "45",
  },
  {
    "type": "cast",
    "value": "number",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "command_type",
    "value": "normal",
  },
  {
    "type": "command",
    "value": "set",
  },
  {
    "type": "reference",
    "value": "y",
  },
  {
    "type": "arrow",
  },
  {
    "type": "reference",
    "value": "user.id",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize function call with mixed arguments 1`] = `
[
  {
    "type": "command_type",
    "value": "call",
  },
  {
    "type": "reference",
    "value": "alert",
  },
  {
    "type": "paren_open",
  },
  {
    "type": "literal",
    "value": "Welcome",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "comma",
  },
  {
    "type": "reference",
    "value": "user.id",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "comma",
  },
  {
    "type": "literal",
    "value": "3000",
  },
  {
    "type": "cast",
    "value": "number",
  },
  {
    "type": "paren_close",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize multiple calls separated by semicolon 1`] = `
[
  {
    "type": "command_type",
    "value": "call",
  },
  {
    "type": "reference",
    "value": "show",
  },
  {
    "type": "paren_open",
  },
  {
    "type": "literal",
    "value": "Saved!",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "paren_close",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "command_type",
    "value": "call",
  },
  {
    "type": "reference",
    "value": "refresh",
  },
  {
    "type": "paren_open",
  },
  {
    "type": "paren_close",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "command_type",
    "value": "call",
  },
  {
    "type": "reference",
    "value": "log",
  },
  {
    "type": "paren_open",
  },
  {
    "type": "literal",
    "value": "done",
  },
  {
    "type": "cast",
    "value": "string",
  },
  {
    "type": "paren_close",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize reference chaining with @scope 1`] = `
[
  {
    "type": "command_type",
    "value": "normal",
  },
  {
    "type": "reference",
    "value": "set",
  },
  {
    "type": "reference",
    "value": "this.user.age",
  },
  {
    "type": "arrow",
  },
  {
    "type": "literal",
    "value": "30",
  },
  {
    "type": "cast",
    "value": "number",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;

exports[`Tokenizer > should tokenize simple function call with no cast 1`] = `
[
  {
    "type": "command_type",
    "value": "call",
  },
  {
    "type": "reference",
    "value": "log",
  },
  {
    "type": "paren_open",
  },
  {
    "type": "literal",
    "value": "debug",
  },
  {
    "type": "paren_close",
  },
  {
    "type": "semicolon",
  },
  {
    "type": "eof",
  },
]
`;
